ddtrace/tracer/span.go

```diff
@@ -169,6 +170,39 @@ func (s *span) setSamplingPriority(priority int, sampler samplernames.SamplerNam
 	s.setSamplingPriorityLocked(priority, sampler, rate)
 }
 
+// setUser sets the span user ID tag as well as some optional user monitoring tags depending on the configuration.
+// The function assumes that the span it is called on is the trace's root span.
+func (s *span) setUser(id string, cfg UserMonitoringConfig) {
+	trace := s.context.trace
+	s.Lock()
+	defer s.Unlock()
+	trace.mu.Lock()
+	defer trace.mu.Unlock()
```

Comment:
This is way too complicated. Red alert 😄 🟥 

Suggestion:
* Rename `s.context.trace.setPropagatingTag` to `s.context.trace.setPropagatingTagLocked`
* Create a new method `s.context.trace.setPropagatingTag` which calls the above and uses the lock.

Same thing, but better distribution of responsibilities. IMHO locks should only be accessed by the owner, not called from outside.

Reply:
Lastly, are we locking things here because @ajgajg1134 says so or because we need to lock? 😄 I'd expect a test to catch the race condition that would've been caused by not locking. Can you please add one?

Reply:
Locking is definitely needed here
```
        tr := newTracer()
	defer tr.Stop()
	t.Run("nested-race", func(t *testing.T) {
		root := tr.newRootSpan("root", "test", "test")

		go func() {
			for i := 0; i < 10000; i++ {
				SetUser(root, "test")
			}
		}()
		go func() {
			for i := 0; i < 10000; i++ {
				tr.StartSpan("test", ChildOf(root.Context())).Finish()
			}
		}()
		root.Finish()
	})
```
This is a race example using only public API, available to customers.

Regarding your suggestion, we will also need a safe method for deletion then, i.e `s.context.trace.setPropagatingTag`

Reply:
Makes sense, perhaps `unsetPropagationTag`? Also, what do you think about adding that test to the PR?

Reply:
Yes, I mean `unsetPropagatingTag` indeed 😄. I added the test as well